crates/rust-analyzer/src/main_loop.rs

```diff
@@ -166,6 +176,19 @@ impl GlobalState {
         Err("client exited without proper shutdown sequence".into())
     }
 
+    fn next_non_lsp_event(&self) -> Option<Event> {
```

Comment:
Maybe we could avoid the `unwrap`s here?

Reply:
You mean something like:
```
    fn next_non_lsp_event(&self) -> Option<Event> {
        select! {
            recv(self.task_pool.receiver) -> task =>
                task.ok().map(Event::Task),

            recv(self.loader.receiver) -> task =>
                task.ok().map(Event::Vfs),

            recv(self.flycheck_receiver) -> task =>
                task.ok().map(Event::Flycheck),
        }
    }
```
?

Reply:
Yeah ^

Reply:
Yeah ^